UX: Detect `fixup!` and `squash!` commits and suggest autosquash instead of failing generic validation

**Description**
When working with a stacked workflow, users frequently use `git commit --fixup` or `git commit --squash` to modify previous commits in the stack. By design, the GHerrit `commit-msg` hook ignores these commits and does not assign them a `gherrit-pr-id` trailer, as they are intended to be transient.

However, if a user attempts to run `git push` before squashing these commits, the `pre-push` hook currently fails with a generic and confusing error message because it expects *every* commit in the range to have a trailer.

**Current Behavior**
In `src/pre_push.rs`, the `collect_commits` function attempts to convert every commit in the range into a `Commit` struct. This parsing logic enforces the presence of the ID:

```rust
// src/pre_push.rs
let gherrit_id = {
    let re = gherrit_pr_id_re();
    let captures = re
        .captures(&message_body)
        .ok_or_else(|| eyre!("Commit {} missing gherrit-pr-id trailer", c.id))?; // <--- FAILS HERE
    captures.get(1).unwrap().as_str().to_string()
};
```

**Resulting Error:** `Error: Commit <hash> missing gherrit-pr-id trailer`

This error implies the tool is broken or the commit is malformed, rather than guiding the user to perform the necessary Git operation.

**Proposed Solution**
Modify the validation logic in `pre_push.rs` to detect standard Git `fixup!` and `squash!` prefixes before checking for the trailer.

If such a commit is found:

1.  **Abort the push** (as we do not want to sync "dirty" history to PRs).
2.  **Display a helpful message:**
    > `[gherrit] [ERROR] Stack contains pending fixup/squash commits.`
    > `[gherrit] Please squash your history before syncing:`
    > `[gherrit]    git rebase -i --autosquash origin/main`

**Implementation Details**

  * Update `Commit::try_from` or `collect_commits` in `src/pre_push.rs`.
  * Check if `message_title` starts with ` fixup!  ` or ` squash!  `.
  * Return a specific error type or message that suggests the autosquash command.

--- a/src/pre_push.rs
+++ b/src/pre_push.rs
@@ -101,7 +101,29 @@ fn collect_commits(repo: &util::Repo) -> Result<Vec<Commit>> {
         .collect::<Result<Vec<_>>>()?;
     commits.reverse();
 
-    commits.into_iter().map(|c| c.try_into()).collect()
+    let remote = repo.default_remote_name();
+    commits
+        .into_iter()
+        .map(|c| -> Result<Commit> {
+            let msg = c.message()?;
+            let title = core::str::from_utf8(msg.title)?;
+
+            if ["fixup!", "squash!", "amend!"]
+                .iter()
+                .any(|p| title.starts_with(p))
+            {
+                // FIXME: Currently, the indent before `git rebase` is not
+                // preserved.
+                bail!(
+                    "Stack contains pending fixup/squash/amend commits.\n\
+                    Please squash your history before syncing:\n\
+                        git rebase -i --autosquash {remote}/{default_branch}",
+                );
+            }
+
+            c.try_into()
+        })
+        .collect()
 }
 
 fn create_gherrit_refs(repo: &util::Repo, commits: Vec<Commit>) -> Result<Vec<Commit>> {

--- a/tests/common/mod.rs
+++ b/tests/common/mod.rs
@@ -172,7 +172,7 @@ fn install_mock_binaries(path: &Path) {
     fs::copy(&gherrit_bin, &gherrit_dst).unwrap();
 }
 
-fn init_git_bare_repo(path: &Path) {
+pub fn init_git_bare_repo(path: &Path) {
     fs::create_dir(path).unwrap();
     run_git_cmd(path, &["init", "--bare"]);
 }

--- a/tests/pre_push_autosquash.rs
+++ b/tests/pre_push_autosquash.rs
@@ -1,0 +1,187 @@
+// Tests for the detection and rejection of autosquash commits (fixup!, squash!,
+// amend!) in the pre-push hook. These tests ensure that GHerrit prevents users
+// from pushing unfinished work to the remote, maintaining a clean commit
+// history.
+
+mod common;
+use common::TestContext;
+use predicates::prelude::*;
+
+// Helper to assert the specific error message format
+fn assert_autosquash_error(
+    output: assert_cmd::assert::Assert,
+    remote: &str,
+    branch: &str,
+) -> assert_cmd::assert::Assert {
+    output
+        .failure()
+        .stderr(predicate::str::contains(
+            "Stack contains pending fixup/squash/amend commits",
+        ))
+        .stderr(predicate::str::contains(format!(
+            "git rebase -i --autosquash {remote}/{branch}",
+        )))
+}
+
+#[test]
+fn test_autosquash_prefixes() {
+    // Verify that the pre-push hook detects and rejects commits with any of the
+    // standard autosquash prefixes: "fixup!", "squash!", and "amend!". This
+    // ensures that even if a user creates these commits manually or via other
+    // tools, GHerrit will catch them before they reach the remote.
+
+    let ctx = TestContext::init_and_install_hooks();
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Initial"]);
+
+    let prefixes = ["fixup!", "squash!", "amend!"];
+
+    for prefix in prefixes {
+        // Create a new branch for each prefix test to ensure clean state
+        let branch_name = format!("feature-{}", prefix.replace("!", ""));
+        ctx.run_git(&["checkout", "-b", &branch_name, "main"]);
+
+        // Normal commit
+        ctx.run_git(&["commit", "--allow-empty", "-m", "Work in progress"]);
+
+        // Prefix commit
+        ctx.run_git(&[
+            "commit",
+            "--allow-empty",
+            "-m",
+            &format!("{} Work in progress", prefix),
+        ]);
+
+        let output = ctx.gherrit().args(["hook", "pre-push"]).assert();
+        assert_autosquash_error(output, "origin", "main");
+    }
+}
+
+#[test]
+fn test_buried_autosquash_commit() {
+    // Test that the hook scans the entire stack of commits being pushed, not
+    // just the check-out tip.
+    //
+    // In this scenario:
+    // - Commit A (Base)
+    // - Commit B (Dirty - fixup! Commit A) <--- Should trigger failure
+    // - Commit C (Clean tip)
+    //
+    // Even though HEAD (Commit C) is clean, the push includes Commit B, so it
+    // must fail.
+    let ctx = TestContext::init_and_install_hooks();
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Initial"]);
+
+    ctx.run_git(&["checkout", "-b", "feature-buried"]);
+
+    // 1. Commit A
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Commit A"]);
+
+    // 2. Commit B (Dirty)
+    ctx.run_git(&["commit", "--allow-empty", "-m", "fixup! Commit A"]);
+
+    // 3. Commit C (Normal)
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Commit C"]);
+
+    let output = ctx.gherrit().args(["hook", "pre-push"]).assert();
+    assert_autosquash_error(output, "origin", "main");
+}
+
+#[test]
+fn test_precedence_over_trailer_check() {
+    // Test the edge case where a commit implies two conflicting states:
+    // 1. It has a valid `gherrit-pr-id` trailer (usually allowing the push).
+    // 2. It has a `fixup!` prefix (absolutely forbidding the push).
+    //
+    // The autosquash check must take precedence. A "fixup" commit is by
+    // definition temporary/incomplete, so it doesn't matter if it has a valid
+    // ID or not.
+
+    let ctx = TestContext::init_and_install_hooks();
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Initial"]);
+    ctx.run_git(&["checkout", "-b", "feature-precedence"]);
+
+    // Create a fixup commit that ALSO has a valid trailer. Normally, a trailer
+    // makes a commit valid for Gherrit. But 'fixup!' should still block it
+    // because it's considered "work in progress/to be squashed".
+    let msg = "fixup! Some feature\n\ngherrit-pr-id: G12345";
+    ctx.run_git(&["commit", "--allow-empty", "-m", msg]);
+
+    let output = ctx.gherrit().args(["hook", "pre-push"]).assert();
+
+    // Must fail with autosquash error, NOT missing trailer error
+    let output = assert_autosquash_error(output, "origin", "main");
+
+    let stderr = std::str::from_utf8(&output.get_output().stderr).unwrap();
+    assert!(
+        !stderr.contains("missing gherrit-pr-id trailer"),
+        "Error should not be about missing trailer. Got: {}",
+        stderr
+    );
+}
+
+#[test]
+fn test_dynamic_remote_and_branch_suggestion() {
+    // Test that the error message correctly suggests the command to run even
+    // when the user has a non-standard configuration.
+    // - Remote: "upstream" (instead of "origin")
+    // - Branch: "master" (instead of "main")
+    //
+    // The error message should recommend:
+    //
+    //   git rebase -i --autosquash upstream/master
+
+    let ctx = TestContext::init();
+
+    // Configure default branch to be 'master' to test dynamic branch name
+    // detection.
+    //
+    // Note: `init_with_repo` usually sets main; this renaming operation
+    // ensures that it is `master` *and* should work even if `init_with_repo`
+    // changes to a different default branch name in the future.
+    //
+    // We also explicitly set `init.defaultBranch` to `master` so that Gherrit's
+    // heuristic (which reads this config) finds `master` even if the user has
+    // a global config setting it to `main`.
+    ctx.run_git(&["config", "init.defaultBranch", "master"]);
+    ctx.run_git(&["branch", "-m", "master"]);
+
+    // Add custom remote 'upstream'. We need to actually point it somewhere
+    // valid for `rev-parse` checks if they happen, but the error message
+    // generation just calls `default_remote_name`. However, `gherrit` manages
+    // `origin` by default. We need to tell it to use `upstream`.
+    let upstream_path = ctx.dir.path().join("upstream.git");
+    common::init_git_bare_repo(&upstream_path);
+    ctx.run_git(&["remote", "add", "upstream", upstream_path.to_str().unwrap()]);
+
+    // Configure gherrit to use upstream
+    ctx.run_git(&["config", "gherrit.remote", "upstream"]);
+
+    // Install hooks manually since we didn't use init_and_install_hooks
+    ctx.install_hooks();
+
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Initial"]);
+    ctx.run_git(&["checkout", "-b", "feature-dynamic"]);
+
+    // Create fixup
+    ctx.run_git(&["commit", "--allow-empty", "-m", "fixup! Work"]);
+
+    let output = ctx.gherrit().args(["hook", "pre-push"]).assert();
+
+    assert_autosquash_error(output, "upstream", "master");
+}
+
+#[test]
+fn test_valid_stack_passes() {
+    // Control group test: standard, valid commits should pass without issues.
+
+    let ctx = TestContext::init_and_install_hooks();
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Initial"]);
+    ctx.run_git(&["checkout", "-b", "feature-valid"]);
+
+    // Stack of 3 normal commits
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Commit A"]);
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Commit B"]);
+    ctx.run_git(&["commit", "--allow-empty", "-m", "Commit C"]);
+
+    ctx.gherrit().args(["hook", "pre-push"]).assert().success();
+}

